Recompose the security overview and built-in authentication content to diataxis doc templates (#30422)
quarkusio/quarkus: docs/src/main/asciidoc/security-proactive-authentication-concept.adoc

////
This document is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
////
[id="security-proactive-authentication-concept"]
= Proactive authentication
include::_attributes.adoc[]
:categories: security,web

Proactive authentication is enabled in Quarkus by default. This means that if an incoming request has a credential then that request will always be authenticated, even if the target page does not require authentication.

[[proactive-authentication]]

Requests with an invalid credential will always be rejected, even when the page is public. 

> If you only want to authenticate when the target page requires authentication, you can change the default behavior.

Reviewer: ```suggestion
You can change the default behavior if you only want to authenticate when the target page requires authentication.
```

To disable proactive authentication in Quarkus, set the following attribute in the `application.properties` configuration file: 

[source,xml,options="nowrap",role="white-space-pre"]
----
`quarkus.http.auth.proactive=false` 
----

> If you disable proactive authentication then the authentication process will only be run when an identity is requested, either because there are security rules that require the user to be authenticated, or due to programmatic access to the current identity.

Reviewer: ```suggestion
If you disable proactive authentication, then the authentication process will only be run when an identity is requested, either because security rules require the user to authenticate or due to programmatic access to the current identity.
```

If proactive authentication is in use, accessing the `SecurityIdentity` is a blocking operation. 

> This is because authentication may not have happened yet, and accessing it may require calls to external systems such as databases that may block.

Reviewer: ```suggestion
This is because authentication may have yet to happen, and accessing it may require calls to external systems, such as databases that may block.
```

> For blocking applications this is no problem, however, if you have disabled authentication in a reactive application this will fail (as you cannot do blocking operations on the IO thread). 

Reviewer: ```suggestion
For blocking applications, this is no problem. However, if you have disabled authentication in a reactive application, this will fail (as you cannot do blocking operations on the IO thread).
```

> To work around this you need to `@Inject` an instance of `io.quarkus.security.identity.CurrentIdentityAssociation`, and call the

Reviewer: ```suggestion
To work around this, you need to `@Inject` an instance of `io.quarkus.security.identity.CurrentIdentityAssociation` and call the `Uni<SecurityIdentity> getDeferredIdentity();` method.
```

> `Uni<SecurityIdentity> getDeferredIdentity();` method. 

Reviewer: ```suggestion
```

    }

}
----

[source,java]
----
import javax.annotation.security.RolesAllowed;
import javax.enterprise.context.ApplicationScoped;

@ApplicationScoped
public class HelloService {

    @RolesAllowed("admin")
    public String sayHello() {
        return "Hello";
    }

}
----

=== How to customize authentication exception responses

You can use JAX-RS `ExceptionMapper` to capture Quarkus Security authentication exceptions such as `io.quarkus.security.AuthenticationFailedException`, for example:

[source,java]
----
package io.quarkus.it.keycloak;

import javax.annotation.Priority;
import javax.ws.rs.Priorities;
import javax.ws.rs.core.Response;
import javax.ws.rs.ext.ExceptionMapper;
import javax.ws.rs.ext.Provider;

import io.quarkus.security.AuthenticationFailedException;

@Provider
@Priority(Priorities.AUTHENTICATION)
public class AuthenticationFailedExceptionMapper implements ExceptionMapper<AuthenticationFailedException> {

    @Context
    UriInfo uriInfo;

    @Override
    public Response toResponse(AuthenticationFailedException exception) {
        return Response.status(401).header("WWW-Authenticate", "Basic realm=\"Quarkus\"").build();
    }
}
----

> CAUTION: Some HTTP authentication mechanisms need to handle authentication exceptions themselves in order to create a correct authentication challenge.

Reviewer: ```suggestion
CAUTION: Some HTTP authentication mechanisms need to handle authentication exceptions themselves to create a correct authentication challenge.
```

For example, `io.quarkus.oidc.runtime.CodeAuthenticationMechanism` which manages OpenId Connect authorization code flow authentication, needs to build a correct redirect URL, cookies, etc.
For that reason, using custom exception mappers to customize authentication exceptions thrown by such mechanisms is not recommended.

> In such cases, a safer way to customize authentication exceptions is to make sure the proactive authentication is not disabled and use Vert.x HTTP route failure handlers, as events come to the handler with the correct response status and headers.

Reviewer: ```suggestion
In such cases, a safer way to customize authentication exceptions is to ensure the proactive authentication is not disabled and use Vert.x HTTP route failure handlers, as events come to the handler with the correct response status and headers.
```
